slidechain/postpegout.go

```diff
@@ -0,0 +1,97 @@
+package slidechain
+
+import (
+	"context"
+	"encoding/json"
+	"math"
+	"time"
+
+	"github.com/chain/txvm/crypto/ed25519"
+	"github.com/chain/txvm/errors"
+	"github.com/chain/txvm/protocol/bc"
+	"github.com/chain/txvm/protocol/txvm"
+	"github.com/chain/txvm/protocol/txvm/op"
+	"github.com/chain/txvm/protocol/txvm/txvmutil"
+	"github.com/stellar/go/xdr"
+)
+
+func (c *Custodian) doPostPegOut(ctx context.Context, assetXDR string, anchor, txid []byte, amount, seqnum int64, peggedOut pegOutState, exporter, temp string, pubkey []byte) error {
+	var asset xdr.Asset
+	err := xdr.SafeUnmarshalBase64(assetXDR, &asset)
+	if err != nil {
+		return errors.Wrap(err, "unmarshaling asset xdr")
+	}
+	assetBytes, err := asset.MarshalBinary()
+	if err != nil {
+		return errors.Wrap(err, "marshaling asset bytes")
+	}
+	assetID := bc.NewHash(txvm.AssetID(importIssuanceSeed[:], assetBytes))
+	ref := pegOut{
+		AssetXDR: assetXDR,
+		Temp:     temp,
+		Seqnum:   seqnum,
+		Exporter: exporter,
+		Amount:   amount,
+		Anchor:   anchor,
+		Pubkey:   pubkey,
+	}
+	refdata, err := json.Marshal(ref)
+	if err != nil {
+		return errors.Wrap(err, "marshaling reference data")
+	}
+	b := new(txvmutil.Builder)
+	b.Tuple(func(contract *txvmutil.TupleBuilder) { // {'C', ...}
+		contract.PushdataByte(txvm.ContractCode)
+		contract.PushdataBytes(exportContract1Seed[:])
+		contract.PushdataBytes(exportContract2Prog)
+		contract.Tuple(func(tup *txvmutil.TupleBuilder) { // {'T', pubkey}
+			tup.PushdataByte(txvm.TupleCode)
+			tup.Tuple(func(pktup *txvmutil.TupleBuilder) {
+				pktup.PushdataBytes(pubkey)
+			})
+		})
+		contract.Tuple(func(tup *txvmutil.TupleBuilder) { // {'V', amount, assetID, anchor}
+			tup.PushdataByte(txvm.ValueCode)
+			tup.PushdataInt64(amount)
+			tup.PushdataBytes(assetID.Bytes())
+			tup.PushdataBytes(anchor)
+		})
+		contract.Tuple(func(tup *txvmutil.TupleBuilder) { // {'S', refdata}
+			tup.PushdataByte(txvm.BytesCode)
+			tup.PushdataBytes(refdata)
+		})
+	})
+	b.PushdataInt64(int64(peggedOut)).Op(op.Put)                        // con stack: snapshot; arg stack: selector
```

Comment:
The selector for phase 2 of the export contract has to be zero (to refund the exported funds) or non-zero (to retire them). But AFAICT `peggedOut` is always either `pegOutOK` (1) or `pegOutFail` (3), so this will always retire.